## Comprehensive analysis of the evolutionary conservation of DNA methyaltion
[![Github](https://img.shields.io/badge/-Github-000?style=flat&logo=Github&logoColor=white)](https://github.com/functionalepigenomics)
[![Gmail](https://img.shields.io/badge/-Gmail-c14438?style=flat&logo=Gmail&logoColor=white)](mailto:dzheng.th@gmail.com)

This repository contains codes codes to reproduce the results and figures published in the paper.

"Genetic Basis, Quantitative Nature, and Functional Relevance of Evolutionarily Conserved DNA Methylation"

The probe list of methylation-conserved CpGs (MCCs), their DNAm variability across individuals in 42 tissues, and DNAm associations with 152 specific phenotypes at the CpG-type level are available for download via a user-friendly R Shiny web interface (QNMEC [Quantitative Nature of Methylation under Evolutionary Conservation]; https://epigenome.shinyapps.io/QNMEC/).

### :computer: Repository Structure
The repository is organized as follows:

- MCC: This folder contains scripts for the identification of MCCs.

- Figures: This folder contains codes for generating the main figures.

- Methyaltion_variability: This folder contains scripts related to analyzing DNA methyaltion variability across individuals in 42 tissues.

- mQTL: This folder stores codes for the identification of methylation quantitative trait loci (mQTLs).

- EWAS: This folder contains scripts for the epigenome-wide association study (EWAS).

- Associated-genes: his folder contains codes for the identification of methylation-associated genes.

Please note that some of the code can be easily modified to perform similar analyses, such as EWAS for each cancer. Only one copy is provided here for easier follow-up.

### ðŸ“« Contact me
Email: dzheng.th@gmail.com
